framework/src/base/DumpObjectsProblem.C

```diff
@@ -0,0 +1,347 @@
+DumpObjectsProblem::addInitialCondition(const std::string & type,
+                                        const std::string & name,
+                                        InputParameters parameters)
+{
+  dumpObjectHelper("ICs", type, name, parameters);
+  FEProblem::addInitialCondition(type, name, parameters);
+}
+
+void
+DumpObjectsProblem::addMaterial(const std::string & type,
+                                const std::string & name,
+                                InputParameters parameters)
+{
+  dumpObjectHelper("Materials", type, name, parameters);
+  FEProblem::addMaterial(type, name, parameters);
+}
+
+std::string
+DumpObjectsProblem::deduceNecessaryParameters(const std::string & type,
+                                              const InputParameters & parameters)
+{
+  auto factory_params = stringifyParameters(_factory.getValidParams(type));
+  auto specified_params = stringifyParameters(parameters);
+
+  std::string param_text;
+  for (auto & value_pair : specified_params)
+  {
+    // parameter name
+    const auto & param_name = value_pair.first;
+    const auto & param_value = value_pair.second;
+
+    auto factory_it = factory_params.find(param_name);
+    if (factory_it == factory_params.end() || factory_it->second != param_value)
+      param_text += "    " + param_name + " = " + param_value + '\n';
+  }
+
+  return param_text;
+}
+
+void
+DumpObjectsProblem::dumpObjectHelper(const std::string & system,
+                                     const std::string & type,
+                                     const std::string & name,
+                                     const InputParameters & parameters)
+{
+  auto path = _app.actionWarehouse().getCurrentActionName();
+  auto param_text = deduceNecessaryParameters(type, parameters);
+
+  // clang-format off
+  _generated_syntax[path][system] +=
+        "  [./" + name + "]\n"
+      + "    type = " + type + '\n'
+      +      param_text
+      + "  [../]\n";
+  // clang-format on
+}
+
+void
+DumpObjectsProblem::dumpVariableHelper(const std::string & system,
+                                       const std::string & var_name,
+                                       FEFamily family,
+                                       Order order,
+                                       Real scale_factor,
+                                       const std::set<SubdomainID> * const active_subdomains)
+{
+  auto path = _app.actionWarehouse().getCurrentActionName();
+  std::string param_text;
+
+  if (active_subdomains)
+  {
+    std::string blocks;
+    for (auto & subdomain_id : *active_subdomains)
+    {
+      auto subdomain_name = _mesh.getMesh().subdomain_name(subdomain_id);
+      if (subdomain_name == "")
+        subdomain_name = std::to_string(subdomain_id);
+
+      if (!blocks.empty())
+        blocks += ' ';
+
+      blocks += subdomain_name;
+    }
+
+    if (active_subdomains->size() > 1)
+      blocks = "'" + blocks + "'";
+
+    param_text += "    blocks = " + blocks + '\n';
```

Comment:
Ideally, it would be nice to just be able to plug output directly into input. `blocks` should be `block`